### DISCUSSION TITLE
table默认的导出怎么增加xls或xlsx格式？

### DISCUSSION THREAD

**Discussion by sszhs** [layui/layui#2364] [问答] [answered]
table默认的导出目前是csv，怎么增加xls或xlsx格式的导出？layui-Excel不太会用……

**Comment by sentsim** [comment]
按照 `defaultToolbar` 选项说明接入第三方组件（如 https://docs.sheetjs.com/docs/ ）实现
https://layui.dev/docs/2/table/#options.defaultToolbar

**Comment by sszhs** [comment]
谢谢贤心大神回复，最后经过尝试基本解决了问题，实现方法较复杂，且还存在一个问题，导出数据列名和数据错位，不知道问题出在哪，是否还有更简单的方法？
![QQ拼音截图未命名](https://github.com/user-attachments/assets/b7963441-f3d2-4cfa-8487-12aae21feb14)
![QQ拼音截图未命名2](https://github.com/user-attachments/assets/5a9dff31-bcf8-4a31-a4d7-6bc1c2354357)

**Reply by LSL1618** [reply]
你提取的表头和行数据的顺序没有一一对应，filteredData与headers无法按下标对应，所以导致头和数据错乱！

**Comment by sszhs** [comment]
没有一一对应是因为，表头列名未被正确筛选（未被隐藏的），试了以下方法还是不行，可否帮忙看一下呢？谢谢！
![11111](https://github.com/user-attachments/assets/35951e81-d651-4d79-811a-7f1910cfec05)

**Reply by LSL1618** [reply, marked as answer]
这个Object.values(row)转换出来的行数据数组也有问题，假如工号在表头headers中的第二列，前面的行数据数组你怎么保证工号的值就在第二列，所以就算你过滤了隐藏表头，还是达不到预期效果。建议你改成这样试试：
```
            var filteredCols = cols.filter(col => !col.hidden);
            // 二维数组表头，提取列头
            var headers = [filteredCols.map(col => col.title)];
            // 二维数组行数据，按表头顺序提取列值
            var filteredData = currentPageData.map(row => {
                var rowBody = [];
                filteredCols.forEach((col, c) => {
                    rowBody[c] = row[col.field];
                });
                return rowBody;
            });
            var ws = XLSX.utils.aoa_to_sheet(headers.concat(filteredData));
```
